server/src/domain/audit/audit.service.ts

```diff
@@ -40,4 +65,160 @@ export class AuditService {
       ids: audits.map(({ entityId }) => entityId),
     };
   }
+
+  async getChecksums(dto: FileChecksumDto) {
+    const results: FileChecksumResponseDto[] = [];
+    for (const filename of dto.filenames) {
+      if (!StorageCore.isImmichPath(filename)) {
+        throw new BadRequestException(
+          `Could not get the checksum of ${filename} because the file isn't accessible by Immich`,
+        );
+      }
+
+      const checksum = await this.cryptoRepository.hashFile(filename);
+      results.push({ filename, checksum: checksum.toString('base64') });
+    }
+    return results;
+  }
+
+  async fixItems(items: FileReportItemDto[]) {
+    for (const { entityId: id, pathType, pathValue } of items) {
+      if (!StorageCore.isImmichPath(pathValue)) {
+        throw new BadRequestException(
+          `Could not fix item ${id} with path ${pathValue} because the file isn't accessible by Immich`,
+        );
+      }
+
+      switch (pathType) {
+        case AssetPathType.ENCODED_VIDEO:
+          await this.assetRepository.save({ id, encodedVideoPath: pathValue });
+          break;
+
+        case AssetPathType.JPEG_THUMBNAIL:
+          await this.assetRepository.save({ id, resizePath: pathValue });
+          break;
+
+        case AssetPathType.WEBP_THUMBNAIL:
+          await this.assetRepository.save({ id, webpPath: pathValue });
+          break;
+
+        case AssetPathType.ORIGINAL:
+          await this.assetRepository.save({ id, originalPath: pathValue });
+          break;
+
+        case AssetPathType.SIDECAR:
+          await this.assetRepository.save({ id, sidecarPath: pathValue });
+          break;
+
+        case PersonPathType.FACE:
+          await this.personRepository.update({ id, thumbnailPath: pathValue });
+          break;
+
+        case UserPathType.PROFILE:
+          await this.userRepository.update(id, { profileImagePath: pathValue });
+          break;
+      }
+    }
+  }
+
+  async getFileReport() {
+    const fullPath = (filename: string) => resolve(filename);
+    const hasFile = (items: Set<string>, filename: string) => items.has(filename) || items.has(fullPath(filename));
+    const crawl = async (folder: StorageFolder) =>
+      new Set(await this.storageRepository.crawl({ pathsToCrawl: [StorageCore.getBaseFolder(folder)] }));
+
+    const uploadFiles = await crawl(StorageFolder.UPLOAD);
+    const libraryFiles = await crawl(StorageFolder.LIBRARY);
+    const thumbFiles = await crawl(StorageFolder.THUMBNAILS);
+    const videoFiles = await crawl(StorageFolder.ENCODED_VIDEO);
+    const profileFiles = await crawl(StorageFolder.PROFILE);
+    const allFiles = new Set<string>();
+    for (const list of [libraryFiles, thumbFiles, videoFiles, profileFiles, uploadFiles]) {
+      for (const item of list) {
+        allFiles.add(item);
+      }
+    }
+
+    const track = (filename: string | null) => {
```

Comment:
Can you explain a bit about the purpose of this method?

Reply:
This is the core of figuring out what files are in the filesystem and not in the database. We start with a list of all the scanned files. Then, as we load assets, people, etc. from the database we're able to mark files as "seen". By removing every file that is referenced by an entity in the database we chip away at the list until only untracked files remain.

Reply:
This method specially, is called on every entity and file path we see in the for loops below.